### DISCUSSION TITLE
buildpack env with BP_ or BPL_ or BPI_

### DISCUSSION THREAD

**Discussion by zhoufenqin** [paketo-buildpacks/feedback#38] [General]
I'm confused why some buildpacks envs start with `BP_`  or `BPL_` or `BPI_`? Is any history or some other rules?

Like: https://github.com/paketo-buildpacks/microsoft-openjdk
`BP_JVM_VERSION`
`BPL_DEBUG_ENABLED`
`BPI_JVM_CACERTS`

**Comment by dmikusa** [comment]
There is a convention that we follow naming buildpack env variables:

- `BP_` means the environment variable is applied/used at build time
- `BPL_` means it's applied/used at launch time (hence the `L`). Launch time is often also called runtime. It also means it doesn't impact build in any way, so setting it during a `pack build` would have no value.
- `BPI_` is "internal" and should not be used by end users. It's used when buildpacks need to communicate information from build time to launch/runtime. You can technically set or override these as a user, but we do not consider them a public interface and reserve the right to change them as necessary for buildpack operations and without warning.

Examples:

- `BP_JVM_VERSION` -> user wants to instruct the buildpacks at buildtime which version of Java to install
- `BPL_DEBUG_ENABLED` -> this is never used at build time, it is only evaluated at launch/runtime and toggles on/off Java debug.
- `BPI_JVM_CACERTS` -> this is set by the buildpacks at buildtime and tells the buildpacks information which is read at launch/runtime just before the app starts up.

Hope that helps!

**Reply by fg-j** [reply]
@dmikusa It'd be great if the bit about `BPI_` environment variables could be documented in our docs about other env vars: https://paketo.io/docs/howto/configuration/#environment-variables

**Reply by dmikusa** [reply]
It has been a deliberate decision to not document those. We don't want end users fiddling with them, or thinking that the values are part of the buildpack contract. I wish there was a better way to pass stuff like this between the buildpack & runtime exec.d processes.

I could maybe see documenting the concept/strategy of `BPI_*` in some buildpack developer-facing documentation, but I think it's better if end users just don't know they exist.
